**Modelo Preaprobacion**

*Caso de estudio: se quiere implementar un modelo de preaprobacion de credito a partir de las variables socioeconomicas de un solicitante de credito. Estas variables son las siguientes:
* Edad:indica el numero de años de un solicitante credito.
* ingresos: Indica todos los ingresos que recibe una persona.
* Egresos: Indica todos los gastos mensuales de una persona.
* Monto(EAD): Es el monto  que una persona esta solicitando a traves del credito.


La variable de referencia sera la variable PreAprovacion. Esta variable indica la preaprobacion o la prenegacion de un credito.


1. Se cargan las librerias de trabajo.

In [1]:
import numpy as np
import pandas as pd
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import confusion_matrix

2. se cargan los datos de trabajo


In [18]:
nxl= '/6. SolicitantesCrédito(USD).xlsx'
XDB=pd.read_excel(nxl,sheet_name=0) #control,shify enter

#seleccionamos las variables de trabajo
XDB=XDB[['Edad','Ingresos','Egresos','Monto (EAD)', 'PreApr']]

filas=np.where(~np.isnan(XDB['PreApr']))[0]
XDB=XDB.iloc[filas,:]
XDB=XDB.dropna() #evita datos faltantes


#separa variables de entrada - variables de salida
YDB = XDB['PreApr'] #SALIDA
XDB = XDB[["Edad" , "Ingresos" , "Egresos" , "Monto (EAD)"]] #ENTRADA





3. se implementan el modelo de clasificacion Naive Bayes

In [23]:
mnb=GaussianNB()
mnb.fit(XDB,YDB)

# datos por categoria
ndat= mnb.class_count_
print("el numero de datos por categoria es:\n:" , ndat)
cat=mnb.classes_
print("las categorias son:\n:" , cat)
# determinamos la propiedades estadisticas de las variables
u=mnb.theta_
print("La media de las variables es: \n:",u)
sigma=np.sqrt(mnb.var_)
print("La desviacion estandar de las variables es: \n:",sigma)
print("Los limites superiores las variables es: \n:", u+sigma)
print("Los limites inferiores de las variables es: \n:", u-sigma)

el numero de datos por categoria es:
: [2959. 2883.]
las categorias son:
: [0. 1.]
La media de las variables es: 
: [[  40.98411626  431.42680644  401.54995644  289.41445941]
 [  41.10752688 1029.98947251  900.19312144  781.2021157 ]]
La desviacion estandar de las variables es: 
: [[ 12.21303268 325.79791497 342.77286087 235.95783541]
 [ 12.25106493 472.43342807 525.43337837 469.90038088]]
Los limites superiores las variables es: 
: [[  53.19714893  757.22472141  744.32281731  525.37229482]
 [  53.35859181 1502.42290058 1425.62649981 1251.10249658]]
Los limites inferiores de las variables es: 
: [[ 28.77108358 105.62889147  58.77709557  53.456624  ]
 [ 28.85646196 557.55604444 374.75974307 311.30173482]]


4. se procede con la evaluacion de desempeño del modelo(frente al pronostico de la decision)

In [22]:
ydp=mnb.predict(XDB)
cm=confusion_matrix(YDB,ydp)
print("La matriz de confusion es: \n",cm)

La matriz de confusion es: 
 [[2502  457]
 [ 937 1946]]


**Matriz de confusion**

   0   1
0  VN  FP
1  FN  VP

5. Metricas de desempeño

In [28]:
cm=confusion_matrix(YDB,ydp)
print("La matriz de confusion es: \n",cm)
VN=cm[0,0];FP=cm[0,1]
FN=cm[1,0];VP=cm[1,1]
TD=len(XDB)#TOTAL DE LOS DATOS

#EXACTITUD: MUESTRA EL COMPORTAMIENTO GENERAL DEL MODELO
Ex=(VN+VP)/TD; print("la exactitud es:",Ex)

#tasa de error: porcentaje de equivocacion
Ter=(FN+FP)/TD; print("la tasa de error es:",Ter)

#sensibilida: como se comporta el modelo frente a los preapobados
Sen=VP/(VP+FN); print("la sensibilidad es:",Sen)

#Especificidad: como se comporta el modelo frente a los prenegados
Esp=VN/(VN+FP); print("la especificidad es:",Esp)

#precision: capacidad del modelo para predecir VP
Pr=VP/(VP+FP); print("la precision es:",Pr)

#prediccion negativa: capacidad del modelo para predecir vn
prn=VN/(VN+FN); print("la prediccion negativa es:",prn)

La matriz de confusion es: 
 [[2502  457]
 [ 937 1946]]
la exactitud es: 0.7613830879835672
la tasa de error es: 0.23861691201643273
la sensibilidad es: 0.6749913284772806
la especificidad es: 0.8455559310577898
la precision es: 0.8098210570120683
la prediccion negativa es: 0.7275370747310265


la exactitud debe ser >= a 75% para que sea valido y que me funcione

**Analisis de Resultados**

* De acuerdo con la base de datos se puede identificar un total de 5842 datos, los cuales estan divididos en 2959 en prenegados y 2883 preaprobados. se puede observar igualmente que la categoria de aprobados presento los valore mas altos frente a la variable ededd, ingresos,egresos y monto.

* De la matriz de confusion se puedeobservar que el modelo tiene la tendecnia Aa equivocarse en mayor medida en el pronostico de los valores preaprobados con un total de 937 falsos negativos, frente a un 457 de falsos positivos.

* de acuerdo con las metricas que arroja la matriz de confusion, se puede observar que el modelo alcanzo un aexactitud por encima del 75%, lo que muestra el buen comportamiento del modelo frente al pronostico de los datos en general.(si es menor a 75 se dice que el modelo no es viable para el uso y que si lo van a usar debe ser bajo la responsabilidad de ellos)

* La tasa de error del modelo corresponde a un 25%

* El modelo tiende mas a negar que aprobar creditos,

* se destaca igualmente el buen funcionamiento alcanzado por el modelo en el pronostivo de los verdaderos positivos(precision)el cual alcanzo un porcentaje de 80%en promedio. este porcentaje fue mas bajo frente al pronostico de los creditos prenegados, alcanzandose una prediccion de 72% en promedio

6.evaluamos un solicitante de credito

In [29]:
XI=[35,600,900,300]
ydpi=mnb.predict([XI])
print("La categoria del solicitante es:",ydpi)

La categoria del solicitante es: [0.]


/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


7. creamos un informe con lo que no poseen preaprobacion


In [42]:
nxl = '/6. SolicitantesCrédito(USD).xlsx'
XDB =pd.read_excel(nxl, sheet_name=0)

#seleccionamos las variables de trabajo
XDB = XDB[['Edad','Ingresos','Egresos','Monto (EAD)','PreApr']]

filas_sp = np.where(np.isnan(XDB['PreApr']))[0]
XDBsp = XDB.iloc[filas_sp,:]
XDBsp = XDBsp[['Edad','Ingresos','Egresos','Monto (EAD)' ]]

#procedemos con el pronostico
ydpsp = mnb.predict(XDBsp)

#pegamos el pronostico
TAsq = np.column_stack((filas_sp,XDBsp , ydpsp))
df = pd.DataFrame(TAsq)
df.columns = ['solicitante','Edad','Ingresos','Egresos','Monto (EAD)','PreApr']
display(df)
df.to_excel("ResultadosPreApro.xlsx")

solicitante  Edad    Ingresos      Egresos  Monto (EAD)  PreApr
0             6.0  38.0   383.67775   368.598667   175.847235     0.0
1             7.0  47.0    34.18750    38.034444    26.955871     0.0
2            11.0  46.0  1774.30500  2116.723778   705.434708     1.0
3            13.0  37.0   447.99675   309.206889   518.676028     0.0
4            16.0  22.0   723.13625   794.864667   419.474589     0.0
...           ...   ...         ...          ...          ...     ...
2531       8357.0  24.0   608.29200   337.400222   776.457576     0.0
2532       8358.0  66.0    20.92100    26.856000     8.495159     0.0
2533       8367.0  32.0   283.26575   170.266444   225.300346     0.0
2534       8372.0  35.0   768.28800   847.264444   479.164232     0.0
2535       8377.0  24.0   487.59800   376.183556   424.283721     0.0

[2536 rows x 6 columns]